docs/architecture.md

```diff
@@ -0,0 +1,223 @@
+
+## Entity Component System (ECS)
+
+ECS is a solution to representing a world that consists of many 'things' that may have common properties with each other. This is perfect for many games and simulations, and is more versatile and performant than class-based inheritance. Universe and ActiveScene both use ECS through the [EnTT](https://github.com/skypjack/entt/) library.
+
+Confusingly, ECS is not "a system of entities and components" that many mainstream game engines already follow. ECS is better stated as "Entities, Components, and Systems:" three separate concepts that come together to create an elegant architecture.
+
+* **Entity**: A simple unique integer ID linked to the existance of a 'thing'.
+* **Component**: An (optional) property of an entity. Usually a struct stored in a contiguous container that associates it to an Entity.
+* **System**: A function that operates on entities that have a specific set of components. Best described as 'rules' that determine how some components change over time.
+
+Entities are still just integers, but are treated as **compositions** of different types of components. Components represent what an entity **has or is made of**, which determines what operations can be done to it by systems. The components themselves don't describe any behaviour.
+
+As a simple example: any entity with a *Mass component* is affected by the *Gravity system*, any entity with the *Legs component* affected by the *Walk system*. *Mass* or *Legs* don't directly imply any behaviour; it's the world's job to add meaning to them.
+
+Components **never inherit each other**, and all common properties are split into separate components.
+
+As another example, a Planet can be represented by having a Mass, Radius, Atmosphere, and Solid surface component. A Star can share the same Mass, Radius, and Atmosphere, but with a Luminance and Plasma surface component.
+
+### Encapsulation?
+
+Encapsulation is not needed, since it is 100% certain which components a system will read from and write to. Systems are very much analagous to real-world laws of the physics. The real world won't tell a falling apple to update; it's not the apple's responsibility to move itself. The world doesn't care that the apple is even an apple; gravity will affect any object with mass regardless of what it is.
+
+### Updating the world
+
+Most mainstream game engines update the world per-object like this:
+
+```
+void world_update(world) {
+    for (each object in world) {
+        object.update()
+    }
+}
+
+```
+
+An ECS engine will more resemble this:
+
+```
+void world_update(world) {
+    gravity_system(world);
+    move_system(world);
+}
+
+void gravity_system(world) {
+    for (each entity with mass and velocity) {
+        // accelerate the masses together
+    }
+}
+
+void move_system(world) {
+    for (each entity with velocity and position) {
+        position += velocity
```

Comment:
point out that order of systems is important, as accelleration effects velocity.

Reply:
i mean, you kind of do that below, but it would help readers to see the relationship directly in the example as well.